iocore/utils/PartitionedMap.h

```diff
@@ -0,0 +1,287 @@
+#ifndef _PartitionedMap_h
+#define _PartitionedMap_h
+
+#include "stdint.h"
+#include <cstddef>
+#include <mutex>
+#include <unordered_map>
+#include <atomic>
+#include <thread>
+
+#include "ts/ink_assert.h"
+
+using Mutex     = std::recursive_mutex;
+using LockGuard = std::lock_guard<Mutex>;
+
+inline bool
+threadHasLock(Mutex &m)
+{
+  return pthread_t(m.native_handle()) == pthread_self();
+}
+
+/**
+ * @brief Low overhead implementation of a recursive read write lock.
+ *
+ * This is does deadlock if you try to acquire both a read and then a write
+ * lock on the same thread.
+ *
+ * I copied the API from std::shared_mutex, so this can be used with std::shared_lock,
+ * std::unique_lock, or std::lock_guard.
+ */
+class RWLock : public std::recursive_mutex
+{
+private:
+  std::atomic<uint16_t> m_active_readers;
+
+public:
+  /// write lock
+  void
+  lock()
+  {
+    std::recursive_mutex::lock();
+    for (int i = 0; m_active_readers; i++) {
+      ink_assert(i != 1000); // you can deadlock here if this thread already has a read lock.
+      std::this_thread::yield();
+    }
+  }
+
+  bool
+  try_lock()
+  {
+    if (!std::recursive_mutex::try_lock()) {
+      return false;
+    }
+    if (m_active_readers) {
+      std::recursive_mutex::unlock();
+      return false;
+    }
+    return true;
+  }
+
+  /// write unlock
+  void
+  unlock()
+  {
+    std::recursive_mutex::unlock();
+  }
+
+  /// read lock
+  void
+  lock_shared()
+  {
+    std::recursive_mutex::lock();
+    m_active_readers++;
+    std::recursive_mutex::unlock();
+  }
+
+  /// read unlock
+  void
+  unlock_shared()
+  {
+    m_active_readers--;
+  }
+
+  /// has write lock
+  bool
+  has_lock()
+  {
+    return threadHasLock(*this) && m_active_readers == 0;
+  }
+
+#if DEBUG
+  /// has read lock, ONLY use this for asserts
+  bool
+  has_lock_shared()
+  {
+    return native_handle() == native_handle_type() && m_active_readers;
+  }
+#endif
+};
+
+/// Intended to make datasets thread safe by assigning locks to partitions of data.
+/** Allocates a fixed number of locks and retrives them with a hash. */
+template <typename Mutex_t> struct LockPool {
+  using Index_t = uint8_t;
+
+  LockPool(size_t num_locks) : m_size(num_locks), m_mutex(new Mutex_t[num_locks]) {}
+
+  Index_t
+  getIndex(size_t key_hash) const
+  {
+    return key_hash % m_size;
+  }
+
+  Mutex_t &
+  getMutex(Index_t index) const
+  {
+    return m_mutex[index];
+  }
+
+  size_t
+  size() const
+  {
+    return m_size;
+  }
+
+private:
+  const size_t m_size = 0;
+  Mutex_t *m_mutex;
```

Comment:
Most of the time, instead of declaring a raw pointer, you should declare a `std::unique_ptr<>`. That handles clean up for you and also prevents unwanted copy / assignment methods from working.

Reply:
Actually, why isn't this a `std::vector<Mutext>`?

Reply:
so I should use vector<unique_ptr<Mutex_t>>. ok.